### PR TITLE
Adjust RPC replication config

### PR DESCRIPTION
**Pull request by wxing1292**
**What changed?**
* Adjust RPC replication configs
* Fix bug on DLQ message generation
* Sort replication tasks by shard ID, task ID when querying DLQ
* Fix bug on ReplicationDlqAckLevel initialization

**Why?**
Fix issues related to NDC RPC stack

**How did you test it?**
Manual tests & existing tests

**Potential risks**
N/A

### CODE REVIEW

**Comment by wxing1292** [comment]
NOTE: this PR depends on #1037, do not review until #1037 is landed

File: service/history/configs/config.go [hunk 1]
```diff
@@ -204,19 +204,22 @@ type Config struct {
...
-	ReplicationTaskProcessorStartWait                  dynamicconfig.DurationPropertyFnWithShardIDFilter
-	ReplicationTaskProcessorStartWaitJitterCoefficient dynamicconfig.FloatPropertyFnWithShardIDFilter
-	ReplicationTaskProcessorHostQPS                    dynamicconfig.FloatPropertyFn
-	ReplicationTaskProcessorShardQPS                   dynamicconfig.FloatPropertyFn
+	ReplicationTaskFetcherParallelism                    dynamicconfig.IntPropertyFn
+	ReplicationTaskFetcherAggregationInterval            dynamicconfig.DurationPropertyFn
+	ReplicationTaskFetcherTimerJitterCoefficient         dynamicconfig.FloatPropertyFn
+	ReplicationTaskFetcherErrorRetryWait                 dynamicconfig.DurationPropertyFn
+	ReplicationTaskProcessorErrorRetryWait               dynamicconfig.DurationPropertyFnWithShardIDFilter
+	ReplicationTaskProcessorErrorRetryBackoffCoefficient dynamicconfig.FloatPropertyFnWithShardIDFilter
```
Comment (samarabbas) [inline service/history/configs/config.go:212]: I'm surprised we have a shardIDFilter.  What is the use case for configuring this for one specific shard?

File: service/history/configs/config.go [hunk 2]
```diff
@@ -332,7 +335,7 @@ func NewConfig(dc *dynamicconfig.Collection, numberOfShards int32, isAdvancedVis
 		ReplicationTaskProcessorStartWait:                      dc.GetDurationPropertyFilteredByShardID(dynamicconfig.ReplicationTaskProcessorStartWait, 5*time.Second),
 		ReplicationTaskProcessorStartWaitJitterCoefficient:     dc.GetFloat64PropertyFilteredByShardID(dynamicconfig.ReplicationTaskProcessorStartWaitJitterCoefficient, 0.9),
 		ReplicationTaskProcessorHostQPS:                        dc.GetFloat64Property(dynamicconfig.ReplicationTaskProcessorHostQPS, 1500),
-		ReplicationTaskProcessorShardQPS:                       dc.GetFloat64Property(dynamicconfig.ReplicationTaskProcessorShardQPS, 5),
+		ReplicationTaskProcessorShardQPS:                       dc.GetFloat64Property(dynamicconfig.ReplicationTaskProcessorShardQPS, 30),
```
Comment (samarabbas) [inline service/history/configs/config.go:338]: These per shard rate limits are quite dangerous as it heavily depends on cluster configuration and super hard to come up with a reasonable default.  We should try to convert all per shard config knobs to something which is more manageable like per host

File: service/history/handler.go [hunk 3]
```diff
@@ -1558,8 +1556,8 @@ func (h *Handler) GetDLQReplicationMessages(ctx context.Context, request *histor
 		}
 	}
 
-	for _, replicationTaskInfos := range taskInfoPerExecution {
-		go handleTaskInfoPerExecution(replicationTaskInfos)
+	for _, replicationTaskInfos := range taskInfoPerShard {
+		go handleTaskInfoPerShard(replicationTaskInfos)
```
Comment (samarabbas) [inline service/history/handler.go:1560]: Probably we should look into capping concurrency for this handling logic.  Number of shards could potentially be large.  Starting a go routine per shard might become a performance issue.  I'm ok to create a separate task to address this.

File: service/history/replicationDLQHandler.go [hunk 4]
```diff
@@ -142,6 +142,13 @@ func (r *replicationDLQHandlerImpl) readMessagesWithAckLevel(
 			ScheduledId:  task.GetScheduledId(),
 		})
 	}
+
+	var replicationTasks []*replicationspb.ReplicationTask
+	pageToken = resp.NextPageToken
```
Comment (samarabbas) [inline service/history/replicationDLQHandler.go:147]: I'm curious why capturing it in a different variable?  why can't you just return resp.NextPageToken like we were doing before?
Also if there is a reason I would rather set pageToken to new value immediately after making the call like on line 130.

File: service/history/replicationTaskProcessor.go [hunk 5]
```diff
@@ -398,7 +400,7 @@ func (p *ReplicationTaskProcessorImpl) convertTaskToDLQTask(
 				TaskId:       replicationTask.GetSourceTaskId(),
 				TaskType:     enumsspb.TASK_TYPE_REPLICATION_HISTORY,
 				FirstEventId: firstEvent.GetEventId(),
-				NextEventId:  lastEvent.GetEventId(),
+				NextEventId:  lastEvent.GetEventId() + 1,
```
Comment (samarabbas) [inline service/history/replicationTaskProcessor.go:403]: Definitely needs a comment.  Looks like `FirstEventId` is inclusive and `NextEventId` is exclusive.  Might not be obvious for new developers.

**Review by samarabbas — APPROVED** [review]

File: service/history/configs/config.go (hunk 2 above)
Comment (wxing1292) [reply service/history/configs/config.go:338]: there is a per host configuration just one line above: `ReplicationTaskProcessorHostQPS`

File: service/history/handler.go (hunk 3 above)
Comment (wxing1292) [reply service/history/handler.go:1560]: long term strategy is to use the grpc streaming functionality, which will greatly reduce the complexity

File: service/history/configs/config.go (hunk 1 above)
Comment (wxing1292) [reply service/history/configs/config.go:212]: the initial implementation comes from cadence repo.

i believe per shard configuration can provide some additional control, i.e. Cassandra (host restart, ring instability, etc) can see lots of error related to LTW. per shard configuration can provide some isolation

File: service/history/replicationDLQHandler.go (hunk 4 above)
Comment (wxing1292) [reply service/history/replicationDLQHandler.go:147]: no specific reason, just easy to read

File: service/history/handler.go (hunk 3 above)
Comment (wxing1292) [reply service/history/handler.go:1560]: #1046

File: service/history/configs/config.go (hunk 2 above)
Comment (wxing1292) [reply service/history/configs/config.go:338]: I believe the long term solution (isolation) is to use 2 level throttler, first throttled by shard, then by host

**Pull request merged**